assets/app/scripts/app.js

```diff
@@ -351,3 +351,88 @@ angular
   });
 
 hawtioPluginLoader.addModule('openshiftConsole');
+
+// API Discovery, this runs before the angular app is bootstrapped
+// TODO we want this to be possible with a single request against the API instead of being dependent on the numbers of groups and versions
+hawtioPluginLoader.registerPreBootstrapTask(function(next) {
+  var api = {
+    k8s: {},
+    openshift: {}
+  };
+  var apis = {};
+  var API_DISCOVERY_ERRORS = [];
+  var protocol = window.location.protocol + "//";
+ 
+ // Fetch /api/v1 for legacy k8s resources, we will never bump the version of these legacy apis so fetch version immediately
+  var k8sBaseURL = protocol + window.OPENSHIFT_CONFIG.api.k8s.hostPort + window.OPENSHIFT_CONFIG.api.k8s.prefix;
+  var k8sDeferred = $.get(k8sBaseURL + "/v1")
+  .done(function(data) {
+    api.k8s.v1 = _.indexBy(data.resources, 'name');
+  })
+  .fail(function(data, textStatus, jqXHR) {
+    API_DISCOVERY_ERRORS.push({
+      data: data,
+      textStatus: textStatus,
+      xhr: jqXHR
+    });
+  });
+  
+  // Fetch /oapi/v1 for legacy openshift resources, we will never bump the version of these legacy apis so fetch version immediately
+  var osBaseURL = protocol + window.OPENSHIFT_CONFIG.api.openshift.hostPort + window.OPENSHIFT_CONFIG.api.openshift.prefix;
+  var osDeferred = $.get(osBaseURL + "/v1")
+  .done(function(data) {
+    api.openshift.v1 = _.indexBy(data.resources, 'name');
+  })
+  .fail(function(data, textStatus, jqXHR) {
+    API_DISCOVERY_ERRORS.push({
+      data: data,
+      textStatus: textStatus,
+      xhr: jqXHR
+    });
+  });
+  
+  // Fetch /apis to get the list of groups and versions, then fetch each group/version
+  var apisBaseURL = protocol + window.OPENSHIFT_CONFIG.apis.hostPort + window.OPENSHIFT_CONFIG.apis.prefix;
+  var apisDeferred = $.get(apisBaseURL)
+  .then(function(data) {
+    var apisDeferredVersions = [];
+    _.each(data.groups, function(group) {
+      apis[group.name] = group;
+      var versions = {};
+      _.each(group.versions, function(version) {
```

Comment:
if `version` is a struct, is it odd to be adding in things to the struct returned from the discovery API?